pkg/experiment/metastore/dlq/recovery.go

```diff
@@ -91,62 +91,71 @@ func (r *Recovery) recoverLoop(ctx context.Context) {
 }
 
 func (r *Recovery) recoverTick(ctx context.Context) {
-	err := r.bucket.Iter(ctx, block.DirNameDLQ, func(metaPath string) error {
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
-		return r.recover(ctx, metaPath)
+	err := r.bucket.Iter(ctx, block.DirNameDLQ, func(path string) error {
+		return r.recover(ctx, path)
 	}, objstore.WithRecursiveIter)
 	if err != nil {
-		level.Error(r.logger).Log("msg", "failed to iterate over dlq", "err", err)
+		level.Error(r.logger).Log("msg", "failed to recover block metadata", "err", err)
 	}
 }
 
-func (r *Recovery) recover(ctx context.Context, metaPath string) error {
-	fields := strings.Split(metaPath, "/")
-	if len(fields) != 5 {
-		r.logger.Log("msg", "unexpected path", "path", metaPath)
+func (r *Recovery) recover(ctx context.Context, path string) (err error) {
+	defer func() {
+		if err == nil {
+			// In case we return no error, the block is considered recovered and will be deleted.
+			if delErr := r.bucket.Delete(ctx, path); delErr != nil {
+				level.Warn(r.logger).Log("msg", "failed to delete block metadata", "err", delErr, "path", path)
+			}
+		}
+	}()
+
+	b, err := r.readObject(ctx, path)
+	switch {
+	case err == nil:
+	case errors.Is(err, context.Canceled):
+		return err
+	case r.bucket.IsObjNotFoundErr(err):
+		// This is somewhat opportunistic: the error is likely caused by a competing recovery
+		// process that has already recovered the block, before we've discovered that the
+		// leadership has changed.
+		level.Warn(r.logger).Log("msg", "block metadata not found; skipping", "path", path)
 		return nil
+	default:
+		// This is somewhat opportunistic, as we don't know if the error is transient or not.
+		// we should consider an explicit retry mechanism with backoff and a limit on the
+		// number of attempts.
+		level.Warn(r.logger).Log("msg", "failed to read block metadata; to be retried", "err", err, "path", path)
+		return err
 	}
-	sshard := fields[1]
-	ulid := fields[3]
-	meta, err := r.get(ctx, metaPath)
-	if err != nil {
-		level.Error(r.logger).Log("msg", "failed to get block meta", "err", err, "path", metaPath)
+
+	var meta metastorev1.BlockMeta
+	if err = meta.UnmarshalVT(b); err != nil {
+		level.Error(r.logger).Log("msg", "invalid block metadata; skipping", "err", err, "path", path)
 		return nil
 	}
-	shard, _ := strconv.ParseUint(sshard, 10, 64)
-	if ulid != meta.Id || meta.Shard != uint32(shard) {
-		level.Error(r.logger).Log("msg", "unexpected block meta", "path", metaPath, "meta", fmt.Sprintf("%+v", meta))
+
+	switch _, err = r.metastore.AddRecoveredBlock(ctx, &metastorev1.AddBlockRequest{Block: &meta}); {
+	case err == nil:
 		return nil
-	}
```

Comment:
Removing it from here because this is not a concern of the recovery process – as long as the object can be unmarshaled and the service accepts it, it's a valid metadata entry